README.md

```diff
@@ -1,16 +1,89 @@
 # Namespaced Merkle Tree (NMT)
+![Go version](https://img.shields.io/badge/go-1.14-blue.svg)
+[![API Reference](https://camo.githubusercontent.com/915b7be44ada53c290eb157634330494ebe3e30a/68747470733a2f2f676f646f632e6f72672f6769746875622e636f6d2f676f6c616e672f6764646f3f7374617475732e737667)](https://pkg.go.dev/github.com/lazyledger/nmt)
+![golangci-lint](https://github.com/lazyledger/nmt/workflows/golangci-lint/badge.svg?branch=master)
+![Go](https://github.com/lazyledger/nmt/workflows/Go/badge.svg)
+![codecov.io](https://codecov.io/github/lazyledger/nmt/coverage.svg?branch=master)
 
-This repository contains (or rather will contain) a namespaced Merkle tree library.
+
+A Namespaced Merkle Tree is
+> [...] an ordered Merkle tree that uses a modified hash function
+  so that each node in the tree includes the range of
+  namespaces of the messages in all of the descendants
+  of each node. The leafs in the tree are ordered by the
+  namespace identifiers of the messages.
+  In a namespaced Merkle tree, each non-leaf node in
+  the tree contains the lowest and highest namespace
+  identifiers found in all the leaf nodes that are descendants of the non-leaf node, in addition to the hash of
+  the concatenation of the children of the node. This
+  enables Merkle inclusion proofs to be created that prove to a verifier that all the elements of the tree for
+  a specific namespace have been included in a Merkle
+  inclusion proof.
+
+The concept was first introduced by [@musalbas] in the LazyLedger [academic paper].
+
+## Example
+```go
+// the tree will use this namespace size
+nidSize := uint8(1)
+// the leaves that will be pushed
+data := []namespace.PrefixedData{
+    namespace.PrefixedDataFrom(namespace.ID{0}, []byte("leaf_0")),
+    namespace.PrefixedDataFrom(namespace.ID{0}, []byte("leaf_1")),
+    namespace.PrefixedDataFrom(namespace.ID{1}, []byte("leaf_2")),
+    namespace.PrefixedDataFrom(namespace.ID{1}, []byte("leaf_3"))}
+// use the hasher to set the namespace size as well as 
+// the underlying hash function:
+nmtHasher := defaulthasher.New(nidSize, crypto.SHA256)
```

Comment:
Shouldn't it be defaultHasher for camel case consistency?

Reply:
Ah this is another package, nvm.

Reply:
Doesn't this need to be imported first in the example?

Reply:
I wanted to keep the example minimal and not include the usual boilerplate code (imports + `func main()`). Can change that if you think this is confusing.

Reply:
ok, I'll add in the full `package main+`import`+`func main`.

```go
package main

import (
    "crypto"
    _ "crypto/sha256"

    "github.com/lazyledger/nmt/defaulthasher"
    "github.com/lazyledger/nmt/namespace"
)

func main() {
    // the tree will use this namespace size
    nidSize := uint8(1)
    // the leaves that will be pushed
    data := []namespace.PrefixedData{
        namespace.PrefixedDataFrom(namespace.ID{0}, []byte("leaf_0")),
        namespace.PrefixedDataFrom(namespace.ID{0}, []byte("leaf_1")),
        namespace.PrefixedDataFrom(namespace.ID{1}, []byte("leaf_2")),
        namespace.PrefixedDataFrom(namespace.ID{1}, []byte("leaf_3"))}
    // use the hasher to set the namespace size as well as 
    // the underlying hash function:
    nmtHasher := defaulthasher.New(nidSize, crypto.SHA256)
    tree := nmt.New(nmtHasher)
    for _, d := range data {
        tree.Push(d)
    }
    
    // compute the root
    root := tree.Root()
    // the root's min/max namespace is the min and max namespace of all leaves:
    if root.Min().Equal(namespace.ID{0}) {
        fmt.Printf("Min namespace: %x\n", root.Min())
    }
    if root.Max().Equal(namespace.ID{1}) {
        fmt.Printf("Max namespace: %x\n", root.Max())
    }
    
    // compute proof for namespace 0:
    proof, err := tree.ProveNamespace(namespace.ID{0}); if err != nil {
        panic("unexpected err")
    }
    
    // verify proof using the root and the leaves of namespace 0:
    leafs := []namespace.PrefixedData{
        namespace.PrefixedDataFrom(namespace.ID{0}, []byte("leaf_0")),
        namespace.PrefixedDataFrom(namespace.ID{0}, []byte("leaf_1"))}
    
    if proof.VerifyNamespace(nmtHasher, namespace.ID{0}, leafs, root) {
        fmt.Printf("Successfully verified namespace: %x\n", namespace.ID{0})
    }
    
    if proof.VerifyNamespace(nmtHasher, namespace.ID{2}, leafs, root) {
        panic(fmt.Sprintf("Proof for namespace %x, passed for namespace: %x\n", namespace.ID{0}, namespace.ID{2}))
    }
}
```

Reply:
resolved in https://github.com/lazyledger/nmt/pull/3/commits/129dba0cfe9b1e090b49714fc334537a69ece018